homeassistant/helpers/llm.py

```diff
@@ -0,0 +1,408 @@
+    "dew_point",
+    "cloud_coverage",
+    "persons",
+]
+
+
+def _format_state(hass: HomeAssistant, entity_state: State) -> dict[str, Any]:
+    """Format state for better understanding by a LLM."""
+    entity_registry = er.async_get(hass)
+    entity_state = template.TemplateState(hass, entity_state, collect=False)
+
+    result: dict[str, Any] = {
+        "name": entity_state.name,
+        "entity_id": entity_state.entity_id,
+        "state": entity_state.state_with_unit,
+        "last_changed": dt_util.get_age(entity_state.last_changed) + " ago",
+    }
+
+    if registry_entry := entity_registry.async_get(entity_state.entity_id):
+        if area_name := template.area_name(hass, entity_state.entity_id):
+            result["area"] = area_name
+        if floor_name := template.floor_name(hass, entity_state.entity_id):
+            result["floor"] = floor_name
+        if len(registry_entry.aliases):
+            result["aliases"] = list(registry_entry.aliases)
+
+    attributes: dict[str, Any] = {}
+    for attribute, value in entity_state.attributes.items():
+        if attribute in EXPORTED_ATTRIBUTES:
+            attributes[attribute] = value
+    if attributes:
+        result["attributes"] = attributes
+
+    return result
+
+
+def custom_serializer(schema: Any) -> Any:
+    """Serialize additional types in OpenAPI-compatible format."""
+    from homeassistant.util.color import (  # pylint: disable=import-outside-toplevel
+        color_name_to_rgb,
+    )
+
+    if schema is cv.string:
+        return {"type": "string"}
+
+    if schema is cv.boolean:
+        return {"type": "boolean"}
+
+    if schema is color_name_to_rgb:
+        return {"type": "string"}
+
+    if isinstance(schema, cv.multi_select):
+        return {"enum": schema.options}
+
+    if isinstance(schema, FunctionType):
+        return {}
+
+    return UNSUPPORTED
+
+
+class IntentTool(Tool):
+    """LLM Tool representing an Intent."""
+
+    def __init__(
+        self,
+        intent_type: str,
+        slot_schema: vol.Schema | None = None,
+        description: str | None = None,
+    ) -> None:
+        """Init the class."""
+        if slot_schema is None:
+            slot_schema = vol.Schema({})
+        if description is None:
+            description = f"Execute Home Assistant {intent_type} intent"
+        super().__init__(
+            intent_type,
+            description,
+            convert(slot_schema, custom_serializer=custom_serializer),
+        )
+
+    async def async_call(
+        self,
+        hass: HomeAssistant,
+        platform: str,
+        text_input: str | None,
+        context: Context,
+        language: str | None,
+        agent_id: str | None,
+        conversation_id: str | None,
+        device_id: str | None,
+        assistant: str | None,
+        **kwargs: Any,
+    ) -> Any:
+        """Handle the intent."""
+        slots = {key: {"value": val} for key, val in kwargs.items()}
+
+        if "area" in slots:
```

Comment:
I'm not sure what is happening here with the pre processing and looking up areas and floors. Wouldn't the intents themselves handle all of this?

Reply:
No, intents (the `DynamicServiceIntentHandler` specifically) expect area_id and floor_id. We are doing here a resolving of area_id and floor_id from its name or alias, replicating what is being done in `helpers.intent.async_match_states`.

Reply:
Yes, `DynamicServiceIntentHandler` calls `async_match_states`, but unless the `area` slot contains an area_id, it would fail earlier: https://github.com/home-assistant/core/blob/9b4099950c374ea67f2d2b57a64b54f7c1844289/homeassistant/helpers/intent.py#L519
So we must resolve the area_id from name or alias before that. For the conversation integration, this pre-processing happens during intent recognition. In theory, we could ask the LLM itself to do the same, by providing it the names, id, and aliases for all areas and floors in the system prompt and ask it to specifically refer to floors and areas by their id only, but I thought this way is more reliable.

Reply:
the downside with this approach is that it's exact matching names while LLMs are really good at matching something like "room of the living" to the "living room". 

The good news here is that I think we can start this way, and we can iterate? In a future PR we could add a tool for LLMs to query areas, devices and entities?

Reply:
I'm still missing a basic question of what's happening in the normal intent path that is equivalent of what is happening here.  I continue to search code trying to get whats going on here and it seems like this is equivalent to what is handled by the default conversation agent `_make_slot_lists` and the hassil library `recognize_all`?

Still trying to answer a basic question about what requirements are different between these use cases and looking for a comment explaining the big picture.

Reply:
> it seems like this is equivalent to what is handled by the default conversation agent `_make_slot_lists` and the hassil library `recognize_all`?

Yes. During the intent recognition we create a list of all areas and floors, their IDs, names, and aliases. The result of intent recognition always refers to area or floor by id, if matched. Example: `{"area": {"value": area_id, "text": area_name}}`
While we can provide the list of all areas and floors and their IDs, names, and aliases to the LLM in the system prompt, during my tests I found out that LLMs not always follow the instructions exactly even when explicitly asked, so I have added an additional step to resolve the `area_id` and `floor_id`.

Reply:
I went to the code to double check, and it appears that @synesthesiam has just fixed that with #115445. I will test again, may be we can now remove this part indeed.

Reply:
Removed.